test/Testing.Tests/Substitute/AutoSubstituteTestTests.cs

```diff
@@ -0,0 +1,46 @@
+using System;
+using FluentAssertions;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Moq;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Rocket.Surgery.Extensions.Testing.Tests
+{
+    public class AutoSubstituteTestTests : AutoMockTest
+    {
+        public AutoSubstituteTestTests(ITestOutputHelper outputHelper) : base(outputHelper) { }
+
+        class Impl : AutoMockTest
+        {
+            public Impl(ITestOutputHelper outputHelper) : base(outputHelper)
+            {
+                Logger.LogError("abcd");
+                Logger.LogError("abcd {something}", "somevalue");
+            }
+        }
+
+        [Fact]
+        public void Should_Create_Usable_Logger()
+        {
+            var test = AutoMock.Resolve<Impl>();
+            AutoMock.Mock<ITestOutputHelper>().Verify(x => x.WriteLine(It.IsAny<string>()), Times.AtLeastOnce);
+        }
+
+        [Fact]
+        public void Should_Provide_Values()
+        {
+            var item = AutoMock.Provide(new MyItem());
+            ServiceProvider.GetRequiredService<MyItem>().Should().BeSameAs(item);
+        }
+
+        [Fact]
+        public void Should_Return_Self_For_ServiceProvider()
+        {
+            ServiceProvider.GetRequiredService<IServiceProvider>().Should().Be(ServiceProvider);
+        }
+
+        class MyItem { }
```

Comment:
I've added a few more tests around making sure we don't automock/fake/stub optional constructor properties 
 see: `Should_Not_Fake_Optional_Parameters` and `Should_Populate_Optional_Parameters_When_Provided`